gh-145678: Fix use-after-free in itertools.groupby _grouper_next() (#145679)
python/cpython: Misc/NEWS.d/next/Library/2026-03-09-00-00-00.gh-issue-145678.grouper-uaf.rst

Fix a use-after-free crash in :func:`itertools.groupby` when a user-defined

> ``__eq__`` advanced the parent iterator while the iterator of groups was advanced.

Reviewer: ```suggestion
Fix a use-after-free crash in :func:`itertools.groupby` when the
parent iterator is mutated while the iterator of groups was advanced.
```

> The fix mirrors the protection added in :gh:`143543` for ``groupby_next()``.

Reviewer: ```suggestion
```

Fix a use-after-free in :func:`itertools.groupby`: the internal
``_grouper_next()`` function did not hold strong references to
``gbo->currkey`` and ``igo->tgtkey`` before calling
:c:func:`PyObject_RichCompareBool`, so a re-entrant ``__eq__`` that advanced
the parent iterator (triggering ``groupby_step()`` and ``Py_XSETREF`` on
``currkey``) could free those objects while they were still being compared.

> The fix mirrors the protection added in :gh:`143543` for ``groupby_next()``.

Reviewer 2: ```suggestion
Fix a use-after-free crash in :func:`itertools.groupby` when a user-defined
``__eq__`` advanced the parent iterator while the iterator of groups was advanced.
The fix mirrors the protection added in :gh:`143543` for ``groupby_next()``.
```
